File: tp1/inference.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(){ 
    FILE * fich_base  =  NULL;
    FILE * fich_temp  =  NULL;
    int nbr_propositions, nbr_variables, nbr_clauses, i, but[20], non_but[20];
    char nom_base[20], c;


    printf("Entrez le nom de votre Fichier BC :\n");
    gets(nom_base);
    strcat(nom_base,".cnf");
    fich_base = fopen(nom_base,"r+");

    if(fich_base  ==   NULL) 
        printf("Impossible d'ouvrir BC...\n");
    else{

        fich_temp = fopen("Temp.cnf","rw+");

        if(fich_temp ==  NULL) 
            printf("Impossible de transferer BC \n");
        else{
            fscanf(fich_base, "p cnf %d %d ",&nbr_variables,&nbr_clauses); // on prend la première ligne de la BC
            nbr_clauses +=  1 ; // pour rajouter le non_but
            fprintf (fich_temp,"p cnf %d %d  ""\n", nbr_variables,nbr_clauses); // on met les infos dans le fichier qu'on va traiter
        }
        c = fgetc(fich_base);
        while(c!= EOF){
            c = fgetc(fich_base);
            if(c!=  EOF) fputc(c,fich_temp);
        } // on a recopier le reste de la BC dans le fichier qu'on va traiter

    printf("BC :\n "); //afficher les codes pour l'utilisateur
    printf("1: Na;\t\t 2: Nb;\t\t 3: Nc;\t \n 4: Cea;\t 5: Ceb:\t 6: Cec;\t \n 7: Coa;\t 8: Cob;\t 9: Coc;\t \n 10: Ma;\t 11: Mb;\t 12: Mc;\t\n\n");    
    printf("Donnez le nombre de littereaux : \n"); //prendre les buts à inferer 
    scanf("%d",&nbr_propositions);  


    for(i = 1; i<nbr_propositions+1; i++){

        printf("Entrez le literal %d :  \n",i);
        scanf("%d", &but[i]);
        if(but[i]>-13 && but[i]<13) 
            non_but[i] = but[i]*(-1); //si les codes sot corrects on prend la negation du but (absurde)
        else 
            puts("Erreur, Vous avez entrer un code invalide");
    }


    fprintf(fich_temp, "\n");  //completion du fichier à traiter
    for(i = 1; i<nbr_propositions+1; i++)fprintf(fich_temp,"%d ",non_but[i]);  //Ajout des negations au fichier
    fprintf(fich_temp,"0"); //Ajout des 0 pour marquer la fin

    system("ubcsat -alg saps -i Temp.cnf -solve > results.txt"); //execution solveur
}
fclose(fich_temp);


    int termine = 0; //signaler fin affichage
    FILE *fich  = fopen("results.txt","r+");
    if(fich_base   ==  NULL) printf("Impossible d'acceder aux resultats...\n");
    else{
        char texte[1000];
        while(fgets(texte, 1000, fich) && !termine){
            if(strstr(texte, "# Solution found for -target 0")){

                printf("\n BC U {Non but} est satisfiable. La base infere le but. \n Solution : \n");
                fscanf(fich, "\n");
                while(!strstr(fgets(texte, 1000, fich), "Variables"))
                    printf("%s", texte);
                termine = 1;
            }
        }
        if(termine   ==  0){ //cas de non satisfiablité
        printf("\n BC U {Non but} est non satisfiable. La base n'infere pas le but.\n");
        int j;
        for(j = 1;j<nbr_propositions+1;j++){
            printf("%d ",(-1)*but[j]);
        }
        if(j>2)
            printf("ne peuvent pas etre atteints");
        else 
            printf("ne peut pas etre atteint\n");
        }
    }

    fclose(fich_temp);
}

File: tp1/inference.py
import os
import shutil
import sys

if len(sys.argv) < 2:
    print("Veuillez entrer le nom du fichier cnf en arguments.")
    quit()

if not os.path.exists(sys.argv[1]):
    print("Fichier cnf n'existe pas dant le chemain entré.")
    quit()

ubcnf = sys.argv[1]
newfile = ubcnf
base = os.path.splitext(ubcnf)[0]
os.rename(newfile, base + '.txt')
shutil.copy(base+'.txt', 'copy.txt')
os.rename(base+'.txt', base + '.cnf')
file = open('copy.txt', "r")
line = file.readline().split()

# extracter les information du fichier cnf
nbr_litteraux = int(line[2])
print("Le nombre des litteraux utilisé est: "+str(nbr_litteraux))
print("Le nombre de clauses originaux: "+line[3])
line[3] = str(int(line[3])+1)
print("Le nombre de clauses apres le traitement: "+line[3])
s = " "
newline = s.join(line)
newline = newline+'\n'
file.close()
but = 0


litteraux = {'Na': 1, 'Nb': 2, 'Nc': 3, 'Cea': 4, 'Ceb': 5, 'Cec': 6, 'Ma': 7, 'Mb': 8, 'Mc': 9, 'Coa': 10, 'Cob': 11, 'Coc': 12}
print("Les litteraux sont:", litteraux)
print("\n")

while True:
        litteral = input("Donner le nom de litteral que vous voulez tester: ")
        
        if litteral in litteraux.keys():
             print("litteral de test valide")
             but = litteraux[litteral]
             break
        print("Litteral n'existe pas")

non_but = -1 * but
with open('copy.txt') as f:
    lines = f.readlines()

lines[0] = newline
f1 = open('copy.txt', "w")
f1.writelines(lines)
f1.close()

# ajouter le non de litteral a la fin
f2 = open('copy.txt', "a")
f2.write('\n'+str(non_but)+' 0')
f2.close()

os.rename('copy.txt', 'copy.cnf')
os.system('ubcsat -alg saps -i copy.cnf -solve > tmp.txt')
check = "# No Solution found for"
infere = False
f3 = open('tmp.txt', "r")
for i in f3:
    if check in i:
        infere = True

if (infere == True):
    print("BC infère '" + litteral + "'")
else:
    print("BC n'infère pas '" + litteral + "'")
f3.close()
if os.path.isfile('copy_done.cnf'):
    os.remove('copy_done.cnf')
os.rename('copy.cnf', 'copy_done.cnf')
    


